stadiamaps/ferrostar
# Bring-your-own-map for Web Components

# Background

The Ferrostar Web Component currently initializes the MapLibre map internally within its shadow DOM structure, and provides a post-init configuration callback to allow users to change how it behaves. However, there are some use cases where it would be nice for developer end users to bring their own map (ex: if they _already_ have a MapLibre component in Vue) which is already initialized.

# Implementation overview

This PR aims to address that by adding an optional slot for the map view. Whenever anything is supplied for this slot, the default MapLibre init logic _will not apply_. Otherwise, things will work as before and users should not notice any difference.

# Current status

I'm opening this initially as a draft PR because I am not 100% sure of all the documentation / implications for styling, given that (it would seem) inserting the external map doesn't inherit the styling due to shadow DOM/scoping rules. I assume this is not an issue for a fully componentized setup like Vue (which is, to my understanding, the primary if not exclusive use case).

Opening this now for feedback from interested parties + those more knowledgeable about modern frontend dev than myself. Once we get feedback on whether this is working for interested stakeholders + is the correct approach, I'll finish the documentation and finish the review process.

# How to test

Here's how to test in your own "real" project.

1. Ensure you have a dev environment set up including wasm-pack etc.; Refer to [[The Ferrostar Book](https://stadiamaps.github.io/ferrostar/dev-env-setup.html)]
2. Clone the repo and check out this branch.
3. Within the `web` directory, execute `npm run build`
4. In your own project's `package.json`, change the ref from npm to your local build. For example:

  ```
  "@stadiamaps/ferrostar-webcomponents": "file:../ferrostar/web",
  ```

5. Pass your own map container via the slot AND set the `map` property on your Ferrostar map component after you've initialized it. See [this commit](https://github.com/ianthetechie/ferrostar-web-test/commit/eeeb9e522d23f63220ee427cbc90baace6226666) for a minimal example (not working due to my aforementioned half-baked impl that doesn't have a separate MapLibre component and is missing something with scoped CSS).
